plutus-wallet-api/ledger/Ledger/Typed/Tx.hs

```diff
@@ -0,0 +1,419 @@
+import qualified Language.PlutusCore.Pretty   as PLC
+
+import           Language.PlutusTx
+import qualified Language.PlutusTx.Builtins   as PlutusTx
+import           Language.PlutusTx.Lift       as Lift
+import           Language.PlutusTx.Lift.Class as Lift
+
+import qualified Language.PlutusIR.Compiler   as PIR
+
+import           Data.Coerce
+import           Data.Kind
+import           Data.List                    (foldl')
+import           Data.Proxy
+import qualified Data.Set                     as Set
+
+import           Control.Monad.Except
+
+import qualified Data.Singletons              as Sing
+import qualified Data.Singletons.Prelude.List as Sing
+
+-- | A class that associates a type standing for a connection type with two types, the type of the redeemer
+-- and the data script for that connection type.
+class ScriptType (a :: Type) where
+    -- | The type of the redeemers of this connection type.
+    type RedeemerType a :: Type
+    -- | The type of the data of this connection type.
+    type DataType a :: Type
+
+    -- Defaults
+    type instance RedeemerType a = ()
+    type instance DataType  a = ()
+
+-- | The type of validators for the given connection type.
+type ValidatorType (a :: Type) = DataType a -> RedeemerType a -> Validation.PendingTx -> Bool
+
+-- | The type of a connection.
+data ScriptInstance (a :: Type) where
+    Validator :: ScriptType a => CompiledCode (ValidatorType a) -> ScriptInstance a
+
+-- | Get the address for a script instance.
+scriptAddress :: ScriptInstance a -> Address
+scriptAddress = Tx.scriptAddress . validatorScript
+
+-- | Get the validator script for a script instance.
+validatorScript :: ScriptInstance a -> ValidatorScript
+validatorScript (Validator vc) = ValidatorScript $ fromCompiledCode vc
+
+-- | Defunctionalized symbol for use with 'Apply'.
+data SealedDataTypeSym :: Type Sing.~> Type
+type instance Sing.Apply SealedDataTypeSym a = PlutusTx.Sealed (HashedDataScript (DataType a))
+-- | Defunctionalized symbol for use with 'Apply'.
+data DataTypeSym :: Type Sing.~> Type
+type instance Sing.Apply DataTypeSym a = DataType a
+
+-- | The type of a redeemer function for the given output types and input type.
+type RedeemerFunctionType (outs :: [Type]) (inn :: Type) = Uncurry (Sing.Map SealedDataTypeSym outs) (RedeemerType inn)
+
+-- | A wrapper around lists. The main difference is that we define singletons for this that only witness the shape, not
+-- the values inside.
+data Shallow (xs :: [a]) = Shallow
+
+data instance Sing.Sing (Shallow xs) where
+    SNil :: Sing.Sing (Shallow '[])
+    SCons :: Sing.Sing (Shallow xs) -> Sing.Sing (Shallow (x ': xs))
+
+instance Sing.SingI (Shallow '[]) where
+    sing = SNil
+
+instance Sing.SingI (Shallow xs) => Sing.SingI (Shallow (x ': xs)) where
+    sing = SCons Sing.sing
+
+ignoreDataScripts
+    :: forall (outs :: [Type]) (inn :: Type)
+    . (All Typeable (Sing.Map SealedDataTypeSym outs))
+    => Sing.Sing (Shallow outs)
```

Comment:
Are you sure you need `singletons` in the end? You can just directly define a shallow singletonized version of `[]` without going through `Sing`. `singletons` are useful when you have a bunch of them (especially when they're nested) or need to convert back and forth between explicit and implicit singletons or just want to reuse existing code, but grepping here for `Sing.` doesn't give me a convincing use case for the particular `singletons` library.

Reply:
You know, I think you're right. That's actually what I started with, and then I thought "hang on, I'm reinventing singletons, I should just use it". But in the end I'm not really able to use the library much so there's not much point.